app/models/tag_implication.rb

```diff
@@ -1,19 +1,71 @@
+# frozen_string_literal: true
+
 class TagImplication < ApplicationRecord
+  # FIXME: subquery in order
+  scope :order_for_listing, lambda {
+    order(Arel.sql('is_pending DESC, (SELECT name FROM tags WHERE id = tag_implications.predicate_id), (SELECT name FROM tags WHERE id = tag_implications.consequent_id)'))
+  }
+
+  scope :search_by_tag_name, lambda { |query|
+    tag_ids = Tag.where('name ILIKE ?', "*#{query}*".to_escaped_for_sql_like).select(:id)
```

Comment:
the original `return if query.blank?` is still needed (should just work for chaining when defined using scope)